### ISSUE TITLE
Decker (C) crashes on undoing pasting a contraption after undoing moving it

### ISSUE THREAD

**Issue by Maffsie** [JohnEarnest/Decker#92]
Ran into this while playing with some of the contraptions in the contraption bazaar.
On Windows and Decker 1.45, create a new deck, then paste the widgets from this contraption: https://itch.io/post/9887819.

<details><summary>Contraption code for ease of reproduction</summary>
<pre><code>
%%WGT0{"w":[{"name":"dialogizer styler1","type":"contraption","size":[289,187],"pos":[49,48],"show":"transparent","def":"dialogizer styler","widgets":{"fontlist":{},"fct":{},"bct":{},"fclr":{"value":"47"},"bclr":{"value":"32"},"tft":{},"tfnt":{},"patpal":{},"traplist":{"row":18},"tbr":{},"brdr":{},"button1":{},"idi":{"value":"p"},"result":{},"sti":{}}}],"d":{"dialogizer styler":{"name":"dialogizer styler","size":[289,187],"margin":[0,0,0,0],"description":"generate style code","script":"on view do\nfontlist.value:select name:key from deck.fonts\nfprv.font:first fontlist.rowvalue\nfprv.text:first fontlist.rowvalue\ntfnt.font:first fontlist.rowvalue\ntfnt.text:first fontlist.rowvalue\ntraplist.value:select name:key from deck.contraptions\nbrdr.text:first traplist.rowvalue\nend","image":"%%IMG2ASEAuwD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wAUIQEA/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wD/AP8A/wDZ","widgets":{"fontlist":{"type":"grid","size":[97,176],"pos":[96,11],"locked":1,"script":"on click row do\n view[]\nend","headers":0,"value":{"name":["body","menu","mono"]},"row":0},"fct":{"type":"field","size":[48,16],"pos":[0,11],"locked":1,"font":"mono","border":0,"align":"right","value":"fcolor:"},"bct":{"type":"field","size":[48,16],"pos":[0,25],"locked":1,"font":"mono","border":0,"align":"right","value":"bcolor:"},"fclr":{"type":"field","size":[49,16],"pos":[48,11],"font":"mono","border":1,"value":"32"},"bclr":{"type":"field","size":[49,16],"pos":[48,26],"font":"mono","border":1,"value":"47"},"tft":{"type":"field","size":[48,16],"pos":[0,40],"locked":1,"font":"mono","border":0,"align":"right","value":"tfont:"},"tfnt":{"type":"field","size":[49,16],"pos":[48,41],"border":1,"value":"body"},"patpal":{"type":"canvas","size":[288,10],"pos":[0,1],"locked":1,"border":0,"image":"%%IMG2ASAACiASARMhESISIxIkEiUSJhInEigSKRIqEisSLBItEi4SIBIBEyERIhIjEiQSJRImEicSKBIpEioSKxIsEi0SLhIgBAEEIAIBBCAEAQQgAQEFIAUBBCEDAQQhAgEEIQQiBAEEIgIBASIHIwQBBCMDAQMjBCQEAQQkAgEBJAclBAEEJQIBBSUDJgQBBCYDAQImBScEAQQnAgEEJwQoBAEBKAUBBCgEKQQBASkFAQMpBSoEAQEqBQEEKgQrBAEBKwUBBCsELAQBASwFAQEsBy0EAQEtBgEDLQQuBAEBLgUBAS4HIAcBASAFAQEgBAEEIAEBASABAQcgAQEEIQYBASEFAQEhBCIHAQEiAgEBIgEBASIFIwcBASMDAQEjBiQHAQEkAgEBJAclBwEBJQYBASUDJgcBASYDAQImBScHAQEnAgEBJwIBAScEKAQBASgBAQEoAwEBKAIBASgEKQQBASkBAQEpBQEBKQUqBAEBKgEBASoGAQEqBCsEAQErAQEBKwYBASsELAQBASwBAQEsAwEBLAEBASwFLQQBAS0BAQEtBAEBLQYuBAEBLgEBAS4DAQEuByAHAQEgBQEBIAQBBCABAQEgAQEHIAEBBCEGAQEhBQEBIQQiBwEBIgIBASIBAQEiBSMHAQEjAwEBIwYkBwEBJAIBASQHJQcBASUGAQElAyYHAQEmAwECJgUnBwEBJwIBBCcEKAQBASgBAQEoAwEBKAIBASgEKQQBASkBAQEpBQEBKQUqBAEBKgEBASoGAQEqBCsEAQErAQEBKwYBASsELAQBASwBAQEsAwEBLAEBASwFLQQBAS0BAQEtBAEBLQYuBAEBLgEBAS4DAQEuByAFAQIgAwEEIAQBBCADAQUgAgEFIQQBAiEEAQIhBSIFAQIiAwEDIgUjBQECIwQBAyMEJAUBAiQDAQQkBCUFAQIlBQECJQQmBQECJgMBBCYEJwUBAicGAQEnBCgEAQMoAwEBKAIBASgEKQQBAykFAQEpAQEBKQMqBAEDKgMBBCoEKwQBAysEAQIrBSwEAQMsAwEDLAUtBAEDLQQBAy0ELgQBAy4DAQQuBCAHAQEgAgEBIAcBBiABAQUgAQEGIQYBASEFAQEhBCIHAQEiBAEBIgUjBwEBIwUBASMEJAcBASQCAQEkAgEBJAQlBwEBJQQBASUFJgcBASYCAQEmAgEBJgQnBwEBJwUBAScEKAYBASgDAQEoAgEBKAQpBgEBKQUBASkBAQEpAyoGAQEqAwEBKgcrBgEBKwYBASsELAYBASwFAQEsBS0GAQEtBgEBLQQuBgEBLgMBAS4CAQEuBCAEAQQgAgEEIAQBBiABAQUgAQEGIQMBBCECAQQhBCIEAQQiBAEBIgUjBAEEIwIBBCMEJAQBBCQCAQQkBCUEAQQlBAEBJQUmBAEEJgIBBCYEJwQBBCcFAQEnBCgGAQEoAwEEKAQpBgEBKQMBBSkDKgYBASoDAQQqBCsGAQErAwEEKwQsBgEBLAUBASwFLQYBAS0DAQQtBC4GAQEuAwEELgQgEgETIREiEiMSJBIlEiYSJxIoEikSKhIrEiwSLRIuEiASARMhESISIxIkEiUSJhInEigSKRIqEisSLBItEi4S","scale":1},"traplist":{"type":"grid","size":[96,176],"pos":[192,11],"locked":1,"script":"on click row do\n view[]\nend","headers":0,"value":{"name":["dialogizer styler"]},"row":2},"tbr":{"type":"field","size":[48,16],"pos":[0,55],"locked":1,"font":"mono","border":0,"align":"right","value":"border:"},"brdr":{"type":"field","size":[49,16],"pos":[48,56],"border":1,"value":"dialogizer styler"},"button1":{"type":"button","size":[96,16],"pos":[0,89],"script":"on click do\nresult.text:\"%s:()\n%s.fcolor:%s\n%s.bcolor:%s\n%s.tfont:%s\n%s.border:deck.contraptions.%s\ndd.style[%s]\" format idi.text,idi.text,fclr.text,idi.text,bclr.text,idi.text,tfnt.text,idi.text,brdr.text,idi.text\nend","text":"generate","style":"rect"},"idi":{"type":"field","size":[49,16],"pos":[48,71],"font":"mono","border":1,"value":"o"},"result":{"type":"field","size":[97,80],"pos":[0,107],"script":"on change val do\n \nend"},"sti":{"type":"field","size":[48,16],"pos":[0,70],"locked":1,"font":"mono","border":0,"align":"right","value":"ID:"}}}}}
</code></pre>
</details>

After pasting the widgets, move the selection around on the page, then undo to move them back to the original position when the widgets were pasted. Then undo again, which should undo pasting the widgets. On my system, Decker crashes at this point reproducibly. Looking at the crash in a debugger, I can see the last instruction before the crash seems to be at `decker + 501CAh` which is `call qword ptr [rbx+50h]`. `rbx` is set to `decker + 1679B0h` so it works out to `call qword ptr [decker + 167A00h]` but that entire block of memory is almost entirely zeroes, causing it to crash with an access violation because it then tries to execute code at `0x0`. I have a few windows minidump files (a meg or two in size) and a WinDbg "time travel" recording (4gb in size). I'll update this issue if I can reproduce it on other platforms.

**Comment by Maffsie** [comment]
Actually it looks like this happens with _any_ contraption pasted into Decker and moved on screen. Here's a minimal example that I can reliably reproduce the crash with:

```
%%WGT0{"w":[{"name":"crasher1","type":"contraption","size":[100,100],"pos":[206,121],"def":"crasher","widgets":{"field1":{}}}],"d":{"crasher":{"name":"crasher","size":[100,100],"margin":[0,0,0,0],"widgets":{"field1":{"type":"field","size":[83,84],"pos":[8,8],"value":"once you paste me, move me once, then undo to put me back, then undo again to crash decker"}}}}}
```

**Comment by Maffsie** [comment]
And I can reproduce it on Mac OS as well - looks like the stack trace from the crash report on Mac is more useful than what Windows'll provide - crash happens inside `is_resizable`.

<details><summary>Relevant crash information from Problem Report</summary>
<pre>
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   ???                           	               0x0 ???
1   app                           	       0x106a9025a is_resizable + 181
2   app                           	       0x106a906ec object_editor + 1144
3   app                           	       0x106a99122 main_view + 3171
4   app                           	       0x106a697f0 tick + 405
5   app                           	       0x106a69646 io_run + 26
6   app                           	       0x106a9aa84 main + 3570
7   app                           	       0x106a30e70 start + 52


Thread 0 crashed with X86 Thread State (64-bit):
  rax: 0x0000000106b384b8  rbx: 0x0000000106b30060  rcx: 0x0000000000000201  rdx: 0x0000000000000000
  rdi: 0x0000000106b30060  rsi: 0x0000000106b384b8  rbp: 0x00007ff7b94cca00  rsp: 0x00007ff7b94cc9d8
   r8: 0x0000000000000064   r9: 0x0000000000000340  r10: 0x0000000000004000  r11: 0xffff80094d677a18
  r12: 0x0000000106a59564  r13: 0x000000130000005e  r14: 0x0000000000000000  r15: 0x0000000106a9c377
  rip: 0x0000000000000000  rfl: 0x0000000000210246  cr2: 0x0000000000000000
  
Logical CPU:     2
Error Code:      0x00000014 (no mapping for user instruction read)
Trap Number:     14


Binary Images:
       0x106a2f000 -        0x106aabfff com.beyondloom.decker (1.45) <058b3fe6-297d-37d2-ae94-b35ea075a4d4> /Users/USER/*/Decker.app/Contents/MacOS/app

{
  "captureTime" : "2024-07-01 13:59:53.4130 +0100",
  "pid" : 8589,
  "cpuType" : "X86-64",
  "roots_installed" : 0,
  "bug_type" : "309",
  "procLaunch" : "2024-07-01 13:59:37.9350 +0100",
  "procStartAbsTime" : 59523624615559,
  "procExitAbsTime" : 59539070920728,
  "procName" : "app",
  "procPath" : "\/Users\/USER\/*\/Decker.app\/Contents\/MacOS\/app",
  "bundleInfo" : {"CFBundleShortVersionString":"1.45","CFBundleIdentifier":"com.beyondloom.decker"},
  "vmRegionInfo" : "0 is not in any region.  Bytes before following region: 4406308864\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      __TEXT                      106a2f000-106aac000    [  500K] r-x\/rwx SM=COW  ...nts\/MacOS\/app",
  "exception" : {"codes":"0x0000000000000001, 0x0000000000000000","rawCodes":[1,0],"type":"EXC_BAD_ACCESS","signal":"SIGSEGV","subtype":"KERN_INVALID_ADDRESS at 0x0000000000000000"},
  "termination" : {"flags":0,"code":11,"namespace":"SIGNAL","indicator":"Segmentation fault: 11","byProc":"exc handler","byPid":8589},
  "vmregioninfo" : "0 is not in any region.  Bytes before following region: 4406308864\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      __TEXT                      106a2f000-106aac000    [  500K] r-x\/rwx SM=COW  ...nts\/MacOS\/app",
</pre>
</details>

**Comment by JohnEarnest** [comment]
Thanks for the report.

I believe I've resolved the problem: destroying objects did not necessarily de-select them, which in certain circumstances could result in an attempt to access the fields of an object which was already garbage-collected.

**Comment by JohnEarnest** [comment]
This fix has been incorporated into release v1.46.
